designs/sam_publish_app_cmd.rst

```diff
@@ -0,0 +1,229 @@
+   * Share the app privately with other AWS accounts using the ``--account-ids`` option
+
+
+#. ``sam package`` command can upload local readme/license files to S3.
+
+#. sample app template generated by ``sam init`` will include the "AWS::ServerlessRepo::Application" section.
+
+
+Out-of-Scope
+------------
+#. Create new application or version if the ``--meta-data`` option is used.
+
+#. Manage application permission separately without publishing/updating the app.
+
+#. Specify type of `application permission`_ in sharing options.
+
+#. Recursively publish nested apps in the template.
+
+#. Run through CI/CD pipeline for the application before publishing.
+
+#. Support downloading an app to edite/republish it.
+
+#. Publish to other repositories besides SAR.
+
+#. Recognize template changes and suggest version number.
+
+#. Parse metadata from template without an AWS::ServerlessRepo::Application section.
+
+.. _application permission: https://docs.aws.amazon.com/serverlessrepo/latest/devguide/access-control-resource-based.html#application-permissions
+
+
+User Experience Walkthrough
+---------------------------
+
+Assuming that customers have the following SAM template:
+
+.. code-block:: yaml
+
+    Metadata:
+        AWS::ServerlessRepo::Application:
+            Name: my-app
+            Description: hello world
+            Author: user1
+            SpdxLicenseId: Apache-2.0
+            LicenseUrl: ./LICENSE.txt
+            ReadmeUrl: ./README.md
+            Labels: ['tests']
+            HomepageUrl: https://github.com/user1/my-app-project
+            SemanticVersion: 1.0.0
+            SourceCodeUrl: https://github.com/user1/my-app-project
+
+    Resources:
+        HelloWorldFunction:
+            Type: AWS::Lambda::Function
+            Properties:
+                CodeUri: s3://bucket/hello-world
+
+Package SAM template
+  Run ``sam package --template-file ./template.yaml --output-template-file packaged.yaml --s3-bucket my-bucket``
+  to upload code artifacts, readme and license files to S3 and generate the packaged template.
+
+Create new application in SAR
+  Run ``sam publish app -t ./packaged.yaml`` to publish a new application named my-app in SAR with the first version
+  created as 1.0.0. The app will be created as private by default.
+
+Create new version of an existing SAR application
+  Modify the existing template, give a different SemanticVersion value, and run ``sam publish app -t ./packaged.yaml``.
+
+Crete application/version and set application permission
+  Run ``sam publish app -t ./packaged.yaml --make-public`` to publish the app and share it publicly. If ``--make-private``
+  option is used, the app will only be visible to the owner. If ``--account-ids <account ids>`` is used, the app will be
+  shared with the provided AWS accounts.
+
+Update the metadata of an exsiting application
+  Run ``sam publish app -t ./packaged.yaml --meta-data --application-id <id>`` to update the application metadata.
```

Comment:
Adding another `sam update app` command might be more confusing because customers have to differentiate between update and publish. The most common use cases are creating applications/versions, therefore I don't feel updating metadata worths a separate command.

Reply:
But the the customer is no longer publishing. I understanding was that publish will create a new app or a new version of an existing app. But from what you just described it can also update an app. Maybe there is some semantic is SAR I don't understand. You have mentioned throughout that this command is for publishing. Updating should be a different command and scope.

Reply:
I'm ok with a `sam update app` command. The concern I have is that `publish app` is an action that has constrained meaning of publishing to a repository, while `update app` can be a genral term to update the Lambda function, update the CloudFormation stack, etc. If we make `sam update app` to be a special command for SAR, we may not be able to expand its meaning in the future.